WordPress/openverse-catalog
# Update database to version 002 (add `license_url`, remove `watermarked` and `ingestion_type`)

## Fixes


Fixes #511 by @obulat

## Description


This PR grew from only backfilling `meta_data.license_url` with computed values to updating the columns.

### Changes to the upstream database structure
We have reviewed the structure of the 'upstream' database, and decided that the following changes would be beneficial:
- `license_url` is a required value, so it's better to lift it to be a separate non-nullable column in the database.
- `watermarked` column is not used much downstream, and it can be moved to be a property of the `meta_data` JSONB column.
- `ingestion_type` column is not used by the API or the frontend, so it should probably be moved to a property on `meta_data`. It probably was important when the data cleanup was not done in the catalog. We are currently only using 'provider APIs', and not using 'commoncrawl'. We might want to add one more value, 'provider data dump', when we use data dumps such as from iNaturalist. 

### Updating the existing database
This PR updates the columns (both TSV and DB) and the DDL that creates the local database.

In production, we will need to first run this one-off DAG that:
- backfills `license_url` by using existing `meta_data.license_url` or by computing the value based on `license` and `license_version`, and then removes `license_url` from the `meta_data`
-  backfills `meta_data.watermarked` and `meta_data.ingestion_type` by copying data from the specific columns.
- reports the data on the columns at the end of it.
After the DAG is run, we can double-check the database, and then manually run the queries dropping `watermarked` and `ingestion_type` columns.

### Adding `license_url`
The DAG in this PR handles backfilling the `license_version` data. It has a testing mode for local data that replaces some items' `meta_data` with `null`, `{}` or `{"some_other_property": "value"}` to ensure that it correctly handles the situation when the `meta_data` is null, an empty jsonb object, or a jsonb object that has some properties, but doesn't have the `license_url`.

I am not sure what the best-performing way of running this query would be, so I added to functions to add `license_url`:
- The first option is to find all items that do not have `license_url`, and then iterate over them one-by-one using an `UPDATE` query to compute the `license_url` based on the row values.
- The second option is to run a single `UPDATE` query per each `license`-`license_version` pair, and update all items with this pair. I though it might be advantageous to compute the `license_url` before each query, and use that to bulk update, instead of running so many individual queries.

Which strategy is better, and do the strategies really differ in performance?

I've added a task to make sample data to make this testable, because local data probably all has correct `license_url` property in `meta_data`. This task is skipped when you trigger DAG normally, but we might remove it altogether before deploying just to be extra careful.

While working on this PR, I also found an error in the computing license URL based on the license and license version where the version 2.0 for `by-nc-nd` was returning a URL with `jp` jurisdiction. I fixed that, and edited the tests to match.

## Testing Instructions

If you don't already have some sample data in your local environment, run a couple of provider DAGs until you see `Wrote 100 lines`, and then mark them successful. This will add some data to your local database.

With the current ingestion using `ImageStore` class, all items have `license_url`. Therefore, we will need to change some items to not have it. To do that, trigger the `add_license_url` DAG with the following config: `{"isTest": true}`. That will replace some well-formed `meta_data` objects with `null`,`{}` or jsonb objects without `license_url`.

The DAG will run and should correctly add the `license_url` property to `meta_data`.

Then, try triggering the DAG without config. The DAG should skip the `make_sample_data` step, and basically do nothing because all data is correct.

You can also try to manually set some `meta_data` to null or remove `license_url` from it, and then run the DAG - it should add the necessary data.

## Checklist


[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
